### ISSUE TITLE
Array of objects not parsed correctly

### ISSUE THREAD

**Issue by lmanzurv** [ljharb/qs#252]
When parsing the URL `?param[][id]=5a53336d48ad1c6cba0e6ae8&param[][id]=5a53336d48ad1c6cba0e6ae7` I get

`
param: [{ 
    id: ['5a53336d48ad1c6cba0e6ae8', '5a53336d48ad1c6cba0e6ae7'] 
}]
` 

instead of

`
param: [
    { id: '5a53336d48ad1c6cba0e6ae8' }, 
    { id: '5a53336d48ad1c6cba0e6ae7' }
]
` 

I am parsing as `qs.parse(queryString, { arrayLimit: 50, ignoreQueryPrefix: true })`.

Am I missing something?

**Comment by ljharb** [comment]
This is expected; because of the `[]`, qs can't know if you want one array or two. Try using indexes: `?param[0][id]=5a53336d48ad1c6cba0e6ae8&param[1][id]=5a53336d48ad1c6cba0e6ae7`.

**Comment by lmanzurv** [comment]
But the structure references an array of *objects*. To me it is very evident that having the same key twice inside an array references 2 different objects, not an array inside the property. In that case, why bother with building an array with just 1 object at all?

I suspect wanting an array of objects where a property of the object has an array would be something like `?param[][id][]=5a53336d48ad1c6cba0e6ae8&param[][id][]=5a53336d48ad1c6cba0e6ae7`, which is clearly not the case.

I also see this is not the first time this issue pops up, which it seems to point that there's something wrong with that logic. See issue https://github.com/ljharb/qs/issues/215#issuecomment-335313445.

**Comment by ljharb** [comment]
`[]` at the end of a key means "items with this key belong to the same array". I don't think there's as wide of a convention between `a[][b]` and `a[][c]` being one array or two.

Can you think of a configuration option we could use that would clearly indicate that choice, and would mesh well with the default?

**Comment by lmanzurv** [comment]
Sure, and that means that saying `param[][id][]=X&param[][id][]=Y` would be hard to know if you refer to `[{id: [X,Y]}]` or `[{id: X}, {id: Y}]`. But since I have `param[][id]=X&param[][id]=Y` where I'm using the same key without the suffixed `[]`, it is safe to assume that they DON'T belong to `[{id: [X,Y]}]`, but in fact are `[{id: X}, {id: Y}]`.

In the case of saying `param[][id][]=X&param[][key][]=Y`, you're right, hard to know if you want `[{id: X}, {key:Y}]` or [{id: X, key: Y}]`, but I think that that is another issue altogether. For this issue, the only solution I would think of is to have an option that would allow you to define a schema that would indicate the desired output structure, though I don't have any suggestions as to how to do it.

**Comment by ljharb** [comment]
An option like that seems reasonable, as long as we could come up with corresponding logic on both parse and stringify so that round trips are possible.

**Comment by jchook** [comment]
Just want to weigh in here that, for example, in PHP:

```php
$x[]['id'] = 5;
$x[]['id'] = 6;

// produces:
// [['id' => 5], ['id' => 6]]
```

I agree with OP and think `x[] = 5` is widely understood to mean `x[x.length] = 5`.

**Comment by ljharb** [comment]
@jchook i'm more concerned with how PHP parses a *query string* that has this syntax, not how its unrelated variable syntax works. What does PHP do with `?x[][id]=5&x[][id]=6`?

**Comment by ljharb** [comment]
Fair enough. https://github.com/ljharb/qs/issues/252#issuecomment-412300241 still stands; if someone wants to submit a PR that adds the option.

**Comment by caasi** [comment]
Is there any update of this issue?

**Comment by ljharb** [comment]
@caasi https://github.com/ljharb/qs/issues/252#issuecomment-449697810

**Comment by caasi** [comment]
I try to draft a parse function:

```javascript
const INIT = 0;
const BEGIN = 1;

function parseQuery([key, value]) {
  let mode = INIT;
  let idx = key.length - 1;
  let c;
  let v = '';
  while (c = key[idx]) {
    switch (mode) {
      case INIT:
        if (c === ']') {
          mode = BEGIN;
        } else {
          v = c + v;
        }
        break;
      case BEGIN:
        if (c === '[') {
          if (v) {
            value = { [v]: value };
            v = '';
          } else {
            value = [value];
          }
          mode = INIT;
        } else {
          v = c + v;
        }
        break;
    }
    idx -= 1;
  }
  switch (mode) {
    case INIT:
      return { [v]: value };
    case BEGIN:
      return { ']': value };
  }
}

function concatQuery(a, b) {
  if (!a) return b;
  if (!b) return a;
  // concat arrays
  if (Array.isArray(a) && Array.isArray(b)) {
    return a.concat(b);
  }
  // concat objects
  let ko = {};
  let ks = Object.keys(a);
  for (let k of ks) {
    ko[k] = true;
  }
  for (let k in b) {
    if (ko[k]) continue;
    ks.push(k);
  }
  let res = {};
  for (let k of ks) {
    res[k] = concatQuery(a[k], b[k]);
  }
  return res;
}

function parse(input) {
  return input.split('&').map(s => s.split('='))
    .map(parseQuery)
    .reduce(concatQuery, {});
}

const inputs = [
  'param[][id]=5a53336d48ad1c6cba0e6ae8&param[][id]=5a53336d48ad1c6cba0e6ae7',
  'param[][id][]=X&param[][id][]=Y',
  'categories[][id]=1&categories[][id]=3&categories[][id]=4',
  'categories[][id]=1&categories[][id]=3&categories[][name]=foo',
  'categories[0][id]=1&categories[2][id]=3&categories[3][name]=4',
  'categories[0][id]=1&categories[2][id]=3&categories[20][name]=foo',
];

for (let input of inputs) {
  console.log(input, '=>', JSON.stringify(parse(input)));
}
```

output:

```
param[][id]=5a53336d48ad1c6cba0e6ae8&param[][id]=5a53336d48ad1c6cba0e6ae7 => {"param":[{"id":"5a53336d48ad1c6cba0e6ae8"},{"id":"5a53336d48ad1c6cba0e6ae7"}]}
param[][id][]=X&param[][id][]=Y => {"param":[{"id":["X"]},{"id":["Y"]}]}
categories[][id]=1&categories[][id]=3&categories[][id]=4 => {"categories":[{"id":"1"},{"id":"3"},{"id":"4"}]}
categories[][id]=1&categories[][id]=3&categories[][name]=foo => {"categories":[{"id":"1"},{"id":"3"},{"name":"foo"}]}
categories[0][id]=1&categories[2][id]=3&categories[3][name]=4 => {"categories":{"0":{"id":"1"},"2":{"id":"3"},"3":{"name":"4"}}}
categories[0][id]=1&categories[2][id]=3&categories[20][name]=foo => {"categories":{"0":{"id":"1"},"2":{"id":"3"},"20":{"name":"foo"}}}
```

But I don't understand how `qs` works. I think I need some help and feedbacks.

**Comment by ljharb** [comment]
@caasi i think the first thing I'd suggest is, open up a draft PR that contains just the tests - ie, what output you expect for both parse and stringify. then, we can look at which ones are failing, and see what would be needed to make them pass.
